### ISSUE TITLE
Fatal Exception if ChromiumWebBrowser in loaded but not visible

### ISSUE THREAD

**Issue by jamespearce2006** [cefsharp/CefSharp#2106]
In version 57 (WPF), if a ChromiumWebBrowser is loaded, but is not yet part of the visual tree, an unhandled exception is thrown:

=======================
```
8/11/2017 9:46:27 AM
Exception Details:  This Visual is not connected to a PresentationSource. (System.InvalidOperationException)

   at System.Windows.Media.Visual.PointToScreen(Point point)
   at CefSharp.Wpf.ChromiumWebBrowser.OnLoaded(Object sender, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
   at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
   at MS.Internal.LoadedOrUnloadedOperation.DoWork()
   at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Interop.HwndTarget.OnResize()
   at System.Windows.Interop.HwndTarget.HandleMessage(WindowMessage msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
```

**Comment by Vyati** [comment]
This bug appeared again in 63.0.1 version

**Comment by squarecircuits** [comment]
Confirmed I get this from 63.0.0 onwards.
dont get the exception with 57.0.0.

I am putting a new ChromiumWebBrowser inside a new tab.

Unhandled Exception: System.InvalidOperationException: This Visual is not connected to a PresentationSource.
   at System.Windows.Media.Visual.PointToScreen(Point point)
   at CefSharp.Wpf.ChromiumWebBrowser.UpdateBrowserScreenLocation() in c:\projects\cefsharp\CefSharp.Wpf\ChromiumWebBrowser.cs:line 1627
   at CefSharp.Wpf.ChromiumWebBrowser.OnLoaded(Object sender, RoutedEventArgs routedEventArgs) in c:\projects\cefsharp\CefSharp.Wpf\ChromiumWebBrowser.cs:line 1774
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
   at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
   at MS.Internal.LoadedOrUnloadedOperation.DoWork()
   at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run(Window window)
   at System.Windows.Application.Run()
   at BigWinHelper.App.Main()

**Comment by merceyz** [comment]
I'm unable to reproduce, are you able to provide an example that causes this issue?

**Comment by squarecircuits** [comment]
my code is [here](https://stackoverflow.com/questions/49190383/cefwebbrowser-into-custom-popup-not-refreshing-properties) on StackOverflow.

the BrowserFactory Method is simply this:         

```
public ChromiumWebBrowser browserFactory(String url) {
            ChromiumWebBrowser aBrowser;
            int count = _tabItems.Count;

            aBrowser = new ChromiumWebBrowser();
            aBrowser.Address = url;

            aBrowser.Name = string.Format("aBrowser{0}", count);
            //missing the url hovering

            aBrowser.LifeSpanHandler = lifeSpanDelegate;
            // register the javascript credential injector
            //aBrowser.FrameLoadEnd += (sender, args) => { this.theModel.InjectCredentials(args.Frame, args.Url); };
            //aBrowser.FrameLoadEnd += (sender, args) => { this.ResetHomeFromLogout(args.Frame, args.Url); };
            aBrowser.LoadHandler = new LoadHandler();
            aBrowser.MenuHandler = new ContextMenuHandler();

            return aBrowser;}
```

and the AddEmptyTabItem is this:

```
 public TabItem AddEmptyTabItem()
        {
            int count = _tabItems.Count; 

            tabDynamic.DataContext = null;
            TabItem tab = new TabItem();
            tab.Header = string.Format("Tab {0}", count); //might be bound to URL
            tab.Name = string.Format("tab{0}", count);
            tab.HeaderTemplate = tabDynamic.FindResource("TabHeader") as DataTemplate;

            _tabItems.Insert(count - 1, tab);
            

            tabDynamic.DataContext = _tabItems;


            return tab;
        }
```

**Comment by amaitland** [comment]
Please fork https://github.com/cefsharp/CefSharp.MinimalExample and provide an example. Thanks.

**Comment by marinkobabic** [comment]
Anwendung: MyApp.exe
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet.
Ausnahmeinformationen: System.InvalidOperationException
bei System.Windows.Media.Visual.PointToScreen(System.Windows.Point)
bei CefSharp.Wpf.ChromiumWebBrowser.UpdateBrowserScreenLocation()
bei System.Windows.EventRoute.InvokeHandlersImpl(System.Object, System.Windows.RoutedEventArgs, Boolean)
bei System.Windows.UIElement.RaiseEventImpl(System.Windows.DependencyObject, System.Windows.RoutedEventArgs)
bei System.Windows.BroadcastEventHelper.BroadcastEvent(System.Windows.DependencyObject, System.Windows.RoutedEvent)
bei System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(System.Object)
bei MS.Internal.LoadedOrUnloadedOperation.DoWork()
bei System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
bei System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
bei System.Windows.Media.MediaContext.RenderMessageHandlerCore(System.Object)
bei System.Windows.Media.MediaContext.RenderMessageHandler(System.Object)
bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
bei System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
bei System.Windows.Threading.DispatcherOperation.InvokeImpl()
bei MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
bei System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
bei System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
bei System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
bei MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
bei System.Windows.Threading.DispatcherOperation.Invoke()
bei System.Windows.Threading.Dispatcher.ProcessQueue()
bei System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
bei MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
bei MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
bei System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
bei System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
bei System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
bei MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
bei MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
bei System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
bei System.Windows.Application.RunDispatcher(System.Object)
bei System.Windows.Application.RunInternal(System.Windows.Window)
bei MyTraderBot.App.Main()

**Comment by marinkobabic** [comment]
Also my case it's related to a TabControl. When running all without tabs I never had this issue.

**Comment by amaitland** [comment]
I prefer not to use `PresentationSource.FromVisual(this)` where possible as it's walks up the tree to find the parent, so could potentially be expensive. It does seem reliable and I nobody has provided a sample that reproduces the problem, so going with this for now. If someone can provide an example that reproduces the problem reliably then we'll look for a better solution (maybe only check if the control is actually visible).

**Comment by Vyati** [comment]
To resolve this problem I just add to OnLoaded method in ChromiumWebBrowser.cs
`
            if(PresentationSource.FromVisual(this) != null)
                UpdateBrowserScreenLocation();
`

**Comment by amaitland** [comment]
This was resolved with #2321
